server/backends/chunkstore/chunkstore.go

```diff
@@ -0,0 +1,413 @@
+	c.DeleteBlob(ctx, blobName)
+	w := c.Writer(ctx, blobName, blockSize, 5*time.Second)
+	if _, err := w.Write(data); err != nil {
+		return err
+	}
+	return w.Close()
+}
+
+func (c *Chunkstore) DeleteBlob(ctx context.Context, blobName string) error {
+	index := uint16(0)
+	for {
+		if exists, err := c.chunkExists(ctx, blobName, index); err != nil {
+			return err
+		} else if !exists {
+			return nil
+		}
+		if err := c.deleteChunk(ctx, blobName, index); err != nil {
+			return err
+		}
+		index++
+	}
+}
+
+func chunkName(blobName string, index uint16) string {
+	return blobName + "_" + fmt.Sprintf("%04x", index)
+}
+
+func (c *Chunkstore) chunkExists(ctx context.Context, blobName string, index uint16) (bool, error) {
+	return c.internalBlobstore.BlobExists(ctx, chunkName(blobName, index))
+}
+
+func (c *Chunkstore) readChunk(ctx context.Context, blobName string, index uint16) ([]byte, error) {
+	return c.internalBlobstore.ReadBlob(ctx, chunkName(blobName, index))
+}
+
+func (c *Chunkstore) writeChunk(ctx context.Context, blobName string, index uint16, data []byte) (int, error) {
+	return c.internalBlobstore.WriteBlob(ctx, chunkName(blobName, index), data)
+}
+
+func (c *Chunkstore) deleteChunk(ctx context.Context, blobName string, index uint16) error {
+	return c.internalBlobstore.DeleteBlob(ctx, chunkName(blobName, index))
+}
+
+type chunkstoreReader struct {
+	ctx        context.Context
+	chunkstore *Chunkstore
+	blobName   string
+	chunk      []byte
+	off        int64
+	chunkOff   int
+	chunkIndex uint16
+	startIndex uint16
+	reverse    bool
+}
+
+func (r *chunkstoreReader) advanceOffset(adv int64) {
+	r.chunkOff += int(adv)
+	r.off += adv
+}
+
+func (r *chunkstoreReader) nextChunkIndex() uint16 {
+	if r.reverse {
+		return r.chunkIndex - 1
+	} else {
+		return r.chunkIndex + 1
+	}
+}
+
+func (r *chunkstoreReader) noChunksRead() bool {
+	return r.chunkIndex == r.startIndex
+}
+
+func (r *chunkstoreReader) copyAndAdvanceOffset(dst, src []byte) int {
+	bytesRead := copy(dst, src)
+	r.advanceOffset(int64(bytesRead))
+	return bytesRead
+}
+
+func (r *chunkstoreReader) copyToReadBuffer(p []byte) int {
+	if r.chunkOff >= len(r.chunk) {
+		return 0
+	}
+	if r.reverse {
+		remainingBytesInChunk := len(r.chunk) - r.chunkOff
+		if len(p) <= remainingBytesInChunk {
+			return r.copyAndAdvanceOffset(p, r.chunk[(remainingBytesInChunk-len(p)):])
+		} else {
+			return r.copyAndAdvanceOffset(p[(len(p)-remainingBytesInChunk):], r.chunk[:remainingBytesInChunk])
+		}
+	} else {
+		return r.copyAndAdvanceOffset(p, r.chunk[r.chunkOff:])
+	}
+}
+
+func (r *chunkstoreReader) nextChunkExists() (bool, error) {
+	return r.chunkstore.chunkExists(r.ctx, r.blobName, r.nextChunkIndex())
+}
+
+func (r *chunkstoreReader) getNextChunk() error {
+	if exists, err := r.nextChunkExists(); err != nil {
+		return err
+	} else if !exists {
+		return fmt.Errorf("Opening %v: %w", chunkName(r.blobName, r.nextChunkIndex()), os.ErrNotExist)
```

Comment:
Can you pick an appropriate util/status error and return those any place we use fmt.Errorfs here? (Prolly status.NotFoundErrorf for this one)

They are handy because they play nicely with gRPC so if any of these errors bubble up uncaught, they'll be displayed/handled correctly in client code. We also have utilities for inspecting these errors and switching on their types which is handy.

Reply:
I actually didn't realize buildbuddy had its own error system until I read this comment; I much prefer this model to the standard golang model where errors rarely contain any state at all and wrapping is skin-deep.